### PR TITLE
feat: run eslint/prettier fix against files as a pre-commit hook with lint-staged

### PR DESCRIPTION
**Pull request by raymondfeng**
Run prettier/eslint fix before `git commit`.

I have seen a lot of lint-related issues in PRs. Hopefully this pre-push
hook can decrease such problems and save us from keep telling contributors
to run `npm run lint:fix`.

A `pre-push` hooks is also added to run `npm run build` to prevent code that does not compile from being pushed to remote.

@frbuceta Thank you for the inspiration at #4330 

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: lint-staged.config.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+const path = require('path');
+
+module.exports = {
+  '*.{js,ts,md}': files => {
+    const cwd = process.cwd();
+    files = files.map(f => path.relative(cwd, f));
+    const commands = [
+      `node packages/build/bin/run-prettier --write ${files.join(' ')}`,
+      `node packages/build/bin/run-eslint --report-unused-disable-directives` +
+        ` --fix --cache ${files.join(' ')}`,
```
Comment (bajtos) [inline lint-staged.config.js:15]: Please note that with TypeScript-powered linter checks, it's not enough to lint the staged files. For example, when the signature of a method inside `@loopback/context` changes (e.g. a `void` function is returning a `Promise` now), we need to lint all packages using that particular method. Many of them will not pass [no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md) check.

File: package.json [hunk 2]
```diff
@@ -77,7 +78,9 @@
   "copyright.owner": "IBM Corp.",
   "husky": {
     "hooks": {
-      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS"
+      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS",
+      "pre-commit": "lint-staged",
```
Comment (bajtos) [inline package.json:82]: Cross-posting from https://github.com/strongloop/loopback-next/pull/4330#issuecomment-575685292:

 In my experience, commit hooks are not very fast, which I find annoying and thus prefer to run checks when I decide on my own. Of course, that's just me, other people may have different preferences.

and from  https://github.com/strongloop/loopback-next/pull/4382#issuecomment-586357588:

My experience with pre-commit hooks is not the best, the hook is often run in situations when I don't want it to be executed (e.g. when performing `git rebase`) and thus slowing me down.

File: package.json [hunk 3]
```diff
@@ -77,7 +78,9 @@
...
+      "pre-push": "npm run build"
```
Comment (bajtos) [inline package.json:83]: This is even worse. 
- `npm run build` takes more than a minute to complete, that's way too long waiting just to be able to push my changes to github. 
- What if I have a work-in-progress branch that does not pass the build but I want the code to be backed up on GitHub and/or open a WIP pull request to discuss the early version? IIUC, this new hook will not allow me to do that.

**Comment by bajtos** [comment]
> I have seen a lot of lint-related issues in PRs. Hopefully this pre-push
> hook can decrease such problems and save us from keep telling contributors
> to run `npm run lint:fix`.

If your main motivation for this pull request is to save typing the comments, then I'd like to offer few alternatives:

1. Use [GitHub Saved Replies](https://help.github.com/en/github/writing-on-github/using-saved-replies) to quickly post a comment using the same text.

2. Move the CI linting step from Travis CI to GitHub actions and post a comment to the pull request if the linting failed. This opens new opportunities - we can distinguish between `eslint` and `prettier` errors and tell the user to run `npm run prettier:fix` if the problem is in formatting, or even fix formatting for them and commit the changes to their feature branch.

**Comment by raymondfeng** [comment]
My primary motivation is to minimize basic issues in PRs that fail CI builds and require manual ping-pongs.

1. To make sure commit messages follow our rules
2. To make sure `npm run lint` passes
3. To make sure `npm run build` passes

I can tolerate edge cases, for example, `lint-staged` misses a few eslint violations as our CI will catch that.

Please note the hooks can be skipped during rebase. See https://github.com/typicode/husky#skip-all-hooks-rebase.

We can also use `-n` or `--no-verify` on `git` commands to skip hooks. See https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--n.

**Comment by bajtos** [comment]
> My primary motivation is to minimize basic issues in PRs that fail CI builds and require manual ping-pongs.

Makes sense. I am afraid pre-commit hooks are not a very reliable solution for that, I have seen pull requests that were failing commit lint checks, despite the fact that we already have a pre-commit hook for that.

In general, I am reluctant to introduce duplicate checks for exceptional situations (a PR that fails CI on linting or build problem), because they negatively impact the usual path where developers run `npm test` before pushing to GitHub.

Maybe we should ask ourselves why are people not running `npm test` often enough? I am guilty too. Back in LB3 days, I was running `npm test` almost all the time because it took less than a minute to complete: full `npm test` in `loopback` takes 36 seconds and that includes Node.js tests, Karma tests in browser and linting, full `npm test` in `strong-remoting` is done in 12 seconds. Compare that with `npm test` in `loopback-next` that can take up to 9 minutes 😱 

Instead of adding pre-commit/push checks forcing developers to suffer these long waits, I am proposing a different approach:

1. Create GitHub Actions to post comments to pull requests so that you don't have to post pings manually. We can even tweak this action to post the comment only to first-time contributors or perhaps only to people that are not maintainers.

2. Invest more time in optimizing our build & lint process. Our goal should be to speed up `npm test` to finish under 1 minute, so that contributors are not disincentivized from running `npm test` often.


> Please note the hooks can be skipped during rebase. See https://github.com/typicode/husky#skip-all-hooks-rebase.
> 
> We can also use `-n` or `--no-verify` on `git` commands to skip hooks. See https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--n.

Thank you for the pointers. Personally, I want to keep `commit-message` hook but skip all other hooks. I don't think that's possible using your proposed hooks and the options listed above. 

If we decide to add new hooks, then let's add all of them to `pre-push`. That way I can create a `git alias` that will run `push --no-verify`, while keeping `git commit` with verification.

**Comment by raymondfeng** [comment]
> If we decide to add new hooks, then let's add all of them to pre-push. That way I can create a git alias that will run push --no-verify, while keeping git commit with verification.

Unfortunately not all checks can/should be performed on  `pre-push`. For example, lint-staged only works with `pre-commit`.

**Comment by InvictusMB** [comment]
+1 for pre-commit hook. It would save me tons of context switching.

I'm not in favor of pre-push check though. Build and test are something CI does well. And if either fails it probably highlights a bigger issue than a missed semicolon and would require a due attention. The primary goal of a hook is to address small nonsensial issues that are time killers otherwise.

Also the full test suite never runs successfully on my Windows machine.

I've made a similar PR #5616 with only a pre-commit and a small prettier tweak.

**Comment by InvictusMB** [comment]
@bajtos One could permanently disable hooks by adding `HUSKY_SKIP_HOOKS=1` to his shell.

**Comment by dougal83** [comment]
> Also the full test suite never runs successfully on my Windows machine.

@InvictusMB Please raise an issue to investigate the issue as it would help to identify sources of friction with the development experience.  Happy to take a look. 👍

**Comment by bajtos** [comment]
> One could permanently disable hooks by adding `HUSKY_SKIP_HOOKS=1` to his shell.

Thank you for the tip. Unfortunately, I don't want to disable all hooks - I want to keep the hook that's checking that my commit message is well formatted.

File: lint-staged.config.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: loopback-next
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+const path = require('path');
+
+module.exports = {
```
Comment (bajtos) [inline lint-staged.config.js:8]:
```suggestion
if (['0', 'false'].includes(process.env.LINT_STAGED)) {
  return;
}

module.exports = {
```

**Comment by raymondfeng** [comment]
Close it as https://github.com/strongloop/loopback-next/pull/5616 has been landed.

**Pull request closed without merging**